### PR TITLE
Mmr query api 

### LINKED ISSUE #913: Api to query latest mmr in filebase db
In order to fetch the latest processed block MMR, using the value from the Postgres DB POI table is not accurate enough, as it only records blocks that have store operation. Therefore, this value should be read from the project Filebase database. This will require: 

- We need to provide an API to query the project filebase database
- After the project is completely synced,  we can insert MMR value to filebased DB as per block indexed. We don't need to wait for the next poi block height to query the POI table. 
- Consider removing reset logic, only verify the last poi with its MMR value against the filebased DB.

### PR DESCRIPTION
**Pull request by jiqiang90**
Closes #913

### CODE REVIEW

**Comment by jiqiang90** [comment]
Todo:

- [ ] still unsure how to validate the blockoffset

File: packages/node/src/meta/mmrQuery.controller.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          error: e.message,
+        },
+        HttpStatus.INTERNAL_SERVER_ERROR,
+      );
+    }
+  }
+
+  @Get()
+  async getLatestMmr(@Query() query) {
+    if (query.latest) {
```
Comment (stwiname) [inline packages/node/src/meta/mmrQuery.controller.ts:39]: What happens if it's not latest?

File: packages/node/src/meta/mmrQuery.controller.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  Query,
+} from '@nestjs/common';
+import { getLogger } from '../utils/logger';
+import { MmrQueryService } from './mmrQuery.service';
+
+const logger = getLogger('mmrs');
+
+@Controller('mmrs')
+export class MmrQueryController {
+  constructor(private mmrService: MmrQueryService) {}
```
Comment (stwiname) [inline packages/node/src/meta/mmrQuery.controller.ts:19]: Whats the reason for not using `MmrService` instead?

File: packages/node/src/indexer/mmr.service.ts [hunk 3]
```diff
@@ -53,42 +48,28 @@ export class MmrService implements OnApplicationShutdown {
...
+    this.blockOffset = blockOffset;
 
     // The file based database current leaf length
     const fileBasedMmrLeafLength = await this.fileBasedMmr.getLeafLength();
     // However, when initialization we pick the previous block for file db and poi mmr validation
     // if mmr leaf length 0 ensure the next block height to be processed min is 1.
     this.nextMmrBlockHeight = fileBasedMmrLeafLength + blockOffset + 1;
     // The first Poi in database with null mmr value.
-    const poiWithoutMmr = await this.getFirstPoiWithoutMmr();
+    // const poiWithoutMmr = await this.getFirstPoiWithoutMmr();
```
Comment (stwiname) [inline packages/node/src/indexer/mmr.service.ts:60]: Can be removed

**Review by stwiname — APPROVED** [review]

File: packages/node/src/meta/mmrQuery.controller.ts (hunk 1 above)
Comment (jiqiang90) [reply packages/node/src/meta/mmrQuery.controller.ts:33]: Thanks, I have make it throw if query not supported

**Comment by jiqiang90** [comment]
This require more test before include in release, as we have added logic for handle empty blocks, better merge after 1.0.0

File: packages/node/src/meta/mmrQuery.controller.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+@Controller('mmrs')
+export class MmrQueryController {
+  constructor(private mmrService: MmrService) {}
+
+  @Get(':blockHeight')
+  async getMmr(@Param() params) {
+    try {
+      // add await, otherwise error been skipped
+      return await this.mmrService.getMmr(params.blockHeight);
```
Comment (ianhe8x) [inline packages/node/src/meta/mmrQuery.controller.ts:25]: I'm not sure what is the default way nestjs serialize a Uint8Array. Can you make the process more explicit?

File: packages/node/src/meta/mmrQuery.controller.ts (hunk 1 above)
Comment (ianhe8x) [reply packages/node/src/meta/mmrQuery.controller.ts:33]: Can you make `/mmrs` return latest mmr  and `/mmrs/xxx` return mmr for xxx height?

File: packages/node/src/indexer/indexer.manager.ts [hunk 5]
```diff
@@ -151,19 +166,13 @@ export class IndexerManager {
...
-                value: blockHeight - 1,
-              },
-              { transaction: tx },
-            );
-            this.setBlockOffset(blockHeight - 1);
-          }
           poiBlockHash = poiBlock.hash;
           await this.storeService.setPoi(poiBlock, { transaction: tx });
           this.poiService.setLatestPoiBlockHash(poiBlockHash);
+          await this.storeService.setMetadataBatch(
```
Comment (ianhe8x) [inline packages/node/src/indexer/indexer.manager.ts:172]: `lastPoiHeight` isn't the height that we have a poi record, but the height that we have mmrRoot calculated.

Comment (ianhe8x) [reply packages/node/src/indexer/indexer.manager.ts:172]: lastPoiHeight is ok to be not realtime, so you won't do too many db writes when you are catching up mmr root calculation to poi table.

File: packages/node/src/indexer/events.ts [hunk 6]
```diff
@@ -36,3 +36,9 @@ export interface NetworkMetadataPayload {
   specName: string;
   genesisHash: string;
 }
+
+export interface MmrPayload {
+  leafIndex: number;
```
Comment (ianhe8x) [inline packages/node/src/indexer/events.ts:41]: Why we need it?

**Comment by ianhe8x** [comment]
I still have concerns that increase wordsize isn't the right way to go, and it increases storage size.

**Review by ianhe8x — APPROVED** [review]
